File: companion/src/Noc.Core/Crypto/Handshake.cs
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace Noc.Core.Crypto;

public enum HandshakeMode : byte { Session = 0, PairQr = 1, PairCode = 2 }

/// <summary>Resultado do handshake do lado do servidor.</summary>
public sealed class HandshakeResult
{
    public required HandshakeMode Mode { get; init; }
    public required byte[] TranscriptHash { get; init; }
    public required SecureChannel Channel { get; init; }
    public required string Sas { get; init; }
    public required byte[] ServerHello { get; init; }
}

public static class Handshake
{
    public const int EphLen = 65;
    public const int NonceLen = 32;
    public const int ClientHelloLen = 1 + 1 + 1 + EphLen + NonceLen;

    public static byte[] EncodePoint(ECParameters p)
    {
        var buf = new byte[EphLen];
        buf[0] = 0x04;
        p.Q.X!.CopyTo(buf, 1);
        p.Q.Y!.CopyTo(buf, 33);
        return buf;
    }

    public static ECDiffieHellmanPublicKey DecodePoint(ReadOnlySpan<byte> point)
    {
        if (point.Length != EphLen || point[0] != 0x04) throw new ProtocolException("bad_point");
        var parameters = new ECParameters
        {
            Curve = ECCurve.NamedCurves.nistP256,
            Q = new ECPoint { X = point.Slice(1, 32).ToArray(), Y = point.Slice(33, 32).ToArray() },
        };
        // valida que o ponto está na curva
        using var tmp = ECDiffieHellman.Create(parameters);
        return tmp.PublicKey;
    }

    public static byte[] TranscriptHash(ReadOnlySpan<byte> clientHello, ReadOnlySpan<byte> serverHelloUnsigned) =>
        SHA256.HashData(Wire.Concat(Wire.LabelTranscript, Wire.Lp(clientHello), Wire.Lp(serverHelloUnsigned)));

    public static (byte[] c2s, byte[] s2c) DeriveKeys(byte[] z, byte[] th)
    {
        var okm = HKDF.DeriveKey(HashAlgorithmName.SHA256, z, 64, th, Wire.InfoKeys);
        return (okm[..32], okm[32..]);
    }

    public static string DeriveSas(byte[] z, byte[] th)
    {
        var b = HKDF.DeriveKey(HashAlgorithmName.SHA256, z, 4, th, Wire.InfoSas);
        return (BinaryPrimitives.ReadUInt32BigEndian(b) % 1_000_000).ToString("D6");
    }

    /// <summary>
    /// Processa o ClientHello e produz o ServerHello (já assinado).
    /// Nonce e chave efêmera podem ser injetados para testes determinísticos.
    /// </summary>
    public static HandshakeResult ServerRespond(
        ReadOnlySpan<byte> clientHello, ECDsa pcKey, byte[] pcSpki,
        ECDiffieHellman? ephemeral = null, byte[]? nonceS = null)
    {
        if (clientHello.Length != ClientHelloLen || clientHello[0] != Wire.FrameClientHello)
            throw new ProtocolException("bad_hello");
        if (clientHello[1] != Wire.ProtocolVersion) throw new ProtocolException("version");
        var mode = clientHello[2];
        if (mode > 2) throw new ProtocolException("bad_mode");
        var ephC = clientHello.Slice(3, EphLen);
        using var peer = DecodePoint(ephC);

        var eph = ephemeral ?? ECDiffieHellman.Create(ECCurve.NamedCurves.nistP256);
        try
        {
            nonceS ??= RandomNumberGenerator.GetBytes(NonceLen);
            var ephS = EncodePoint(eph.ExportParameters(false));
            var sh0 = Wire.Concat([Wire.FrameServerHello], ephS, nonceS, Wire.Lp(pcSpki));
            var th = TranscriptHash(clientHello, sh0);
            var sig = pcKey.SignData(Wire.Concat(Wire.LabelServer, th), HashAlgorithmName.SHA256,
                DSASignatureFormat.Rfc3279DerSequence);
            var serverHello = Wire.Concat(sh0, Wire.Lp(sig));
            var z = eph.DeriveRawSecretAgreement(peer);
            var (c2s, s2c) = DeriveKeys(z, th);
            var sas = DeriveSas(z, th);
            CryptographicOperations.ZeroMemory(z);
            return new HandshakeResult
            {
                Mode = (HandshakeMode)mode,
                TranscriptHash = th,
                Channel = new SecureChannel(sendKey: s2c, recvKey: c2s, sendDir: 2, recvDir: 1),
                Sas = sas,
                ServerHello = serverHello,
            };
        }
        finally
        {
            if (ephemeral is null) eph.Dispose();
        }
    }

    public static bool VerifyClientSignature(byte[] clientSpki, byte[] th, byte[] sig)
    {
        try
        {
            using var key = ECDsa.Create();
            key.ImportSubjectPublicKeyInfo(clientSpki, out _);
            if (key.KeySize != 256) return false;
            return key.VerifyData(Wire.Concat(Wire.LabelClient, th), sig, HashAlgorithmName.SHA256,
                DSASignatureFormat.Rfc3279DerSequence);
        }
        catch (CryptographicException) { return false; }
    }

    public static byte[] PairMac(byte[] secret, byte[] th, byte[] clientSpki) =>
        HMACSHA256.HashData(secret, Wire.Concat(Wire.LabelPair, th, clientSpki));
}

File: android/app/src/main/java/com/noc/app/core/crypto/Handshake.kt
package com.noc.app.core.crypto

import java.math.BigInteger
import java.nio.ByteBuffer
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.SecureRandom
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.KeyAgreement

enum class HandshakeMode(val code: Byte) { SESSION(0), PAIR_QR(1), PAIR_CODE(2) }

/** Quem assina pelo celular. No app é a chave do Android Keystore; nos testes, uma chave em software. */
interface DeviceSigner {
    val publicSpki: ByteArray
    fun sign(data: ByteArray): ByteArray
}

object Ec {
    const val POINT_LEN = 65

    val P256: ECParameterSpec by lazy {
        val params = AlgorithmParameters.getInstance("EC")
        params.init(ECGenParameterSpec("secp256r1"))
        params.getParameterSpec(ECParameterSpec::class.java)
    }

    private fun fixed32(v: BigInteger): ByteArray {
        val raw = v.toByteArray()
        return when {
            raw.size == 32 -> raw
            raw.size > 32 -> raw.copyOfRange(raw.size - 32, raw.size)
            else -> ByteArray(32 - raw.size) + raw
        }
    }

    fun encodePoint(key: ECPublicKey): ByteArray =
        byteArrayOf(0x04) + fixed32(key.w.affineX) + fixed32(key.w.affineY)

    fun decodePoint(point: ByteArray): ECPublicKey {
        if (point.size != POINT_LEN || point[0] != 0x04.toByte()) throw ProtocolException("bad_point")
        val x = BigInteger(1, point.copyOfRange(1, 33))
        val y = BigInteger(1, point.copyOfRange(33, 65))
        return KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(ECPoint(x, y), P256)) as ECPublicKey
    }

    fun publicFromSpki(spki: ByteArray): ECPublicKey =
        KeyFactory.getInstance("EC").generatePublic(X509EncodedKeySpec(spki)) as ECPublicKey

    fun verify(spki: ByteArray, data: ByteArray, derSig: ByteArray): Boolean = try {
        val s = Signature.getInstance("SHA256withECDSA")
        s.initVerify(publicFromSpki(spki))
        s.update(data)
        s.verify(derSig)
    } catch (e: Exception) {
        false
    }
}

/**
 * Lado cliente do handshake Noc v1.
 * Para testes determinísticos, `ephemeralPrivate`/`ephemeralPoint`/`nonce` podem ser injetados.
 */
class HandshakeClient(
    mode: HandshakeMode,
    ephemeralPrivate: PrivateKey? = null,
    ephemeralPoint: ByteArray? = null,
    nonce: ByteArray? = null,
) {
    private val ephPrivate: PrivateKey
    val clientHello: ByteArray

    init {
        val point: ByteArray
        if (ephemeralPrivate != null && ephemeralPoint != null) {
            ephPrivate = ephemeralPrivate
            point = ephemeralPoint
        } else {
            val kpg = KeyPairGenerator.getInstance("EC")
            kpg.initialize(ECGenParameterSpec("secp256r1"))
            val kp = kpg.generateKeyPair()
            ephPrivate = kp.private
            point = Ec.encodePoint(kp.public as ECPublicKey)
        }
        val n = nonce ?: ByteArray(NONCE_LEN).also { SecureRandom().nextBytes(it) }
        clientHello = Wire.concat(
            byteArrayOf(Wire.FRAME_CLIENT_HELLO, Wire.PROTOCOL_VERSION, mode.code), point, n,
        )
    }

    class Completed(
        val pcSpki: ByteArray,
        val transcriptHash: ByteArray,
        val channel: SecureChannel,
        val sas: String,
    )

    fun complete(serverHello: ByteArray): Completed {
        if (serverHello.isNotEmpty() && serverHello[0] == Wire.FRAME_ERROR) {
            throw ProtocolException(String(serverHello, 1, serverHello.size - 1, Charsets.UTF_8))
        }
        if (serverHello.size < 1 + Ec.POINT_LEN + NONCE_LEN + 8 || serverHello[0] != Wire.FRAME_SERVER_HELLO) {
            throw ProtocolException("bad_server_hello")
        }
        val ephS = serverHello.copyOfRange(1, 1 + Ec.POINT_LEN)
        val (pcSpki, afterSpki) = Wire.readLp(serverHello, 1 + Ec.POINT_LEN + NONCE_LEN)
        val (sig, end) = Wire.readLp(serverHello, afterSpki, 256)
        if (end != serverHello.size) throw ProtocolException("bad_server_hello")

        val th = transcriptHash(clientHello, serverHello.copyOfRange(0, afterSpki))
        if (!Ec.verify(pcSpki, Wire.concat(Wire.LABEL_SERVER, th), sig)) throw ProtocolException("server_signature")

        val ka = KeyAgreement.getInstance("ECDH")
        ka.init(ephPrivate)
        ka.doPhase(Ec.decodePoint(ephS), true)
        val z = ka.generateSecret()
        val okm = Wire.hkdf(z, th, Wire.INFO_KEYS, 64)
        val sas = deriveSas(z, th)
        z.fill(0)
        val channel = SecureChannel(
            sendKey = okm.copyOfRange(0, 32), recvKey = okm.copyOfRange(32, 64), sendDir = 1, recvDir = 2,
        )
        okm.fill(0)
        return Completed(pcSpki, th, channel, sas)
    }

    companion object {
        const val NONCE_LEN = 32

        fun transcriptHash(clientHello: ByteArray, serverHelloUnsigned: ByteArray): ByteArray =
            Wire.sha256(Wire.LABEL_TRANSCRIPT, Wire.lp(clientHello), Wire.lp(serverHelloUnsigned))

        fun deriveSas(z: ByteArray, th: ByteArray): String {
            val b = Wire.hkdf(z, th, Wire.INFO_SAS, 4)
            val v = ByteBuffer.wrap(b).int.toLong() and 0xffffffffL
            return "%06d".format(v % 1_000_000)
        }

        fun clientSignatureInput(th: ByteArray) = Wire.concat(Wire.LABEL_CLIENT, th)

        fun pairMac(secret: ByteArray, th: ByteArray, deviceSpki: ByteArray): ByteArray =
            Wire.hmacSha256(secret, Wire.concat(Wire.LABEL_PAIR, th, deviceSpki))
    }
}
